engine-spring/core/src/test/java/org/camunda/bpm/engine/spring/test/transaction/crdb/CrdbTxRetryOnCommitTest.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
+ * under one or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information regarding copyright
+ * ownership. Camunda licenses this file to you under the Apache License,
+ * Version 2.0; you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.camunda.bpm.engine.spring.test.transaction.crdb;
+
+import org.camunda.bpm.engine.CrdbTransactionRetryException;
+import org.camunda.bpm.engine.impl.cfg.ProcessEngineConfigurationImpl;
+import org.camunda.bpm.engine.test.Deployment;
+import org.camunda.bpm.engine.test.ProcessEngineRule;
+import org.junit.Ignore;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+import org.junit.runner.RunWith;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.test.context.ContextConfiguration;
+import org.springframework.test.context.junit4.SpringJUnit4ClassRunner;
+
+/**
+ * This tests simulates a CockroachDB concurrency error on TX commit.
+ */
+@RunWith(SpringJUnit4ClassRunner.class)
+@ContextConfiguration(locations = {"classpath:org/camunda/bpm/engine/spring/test/transaction/" +
+    "CrdbTransactionIntegrationTest-applicationContext.xml"})
+public class CrdbTxRetryOnCommitTest {
+
+  @Rule
+  @Autowired
+  public ProcessEngineRule rule;
+
+  @Rule
+  public ExpectedException thrown = ExpectedException.none();
+
+  @Autowired
+  public ProcessEngineConfigurationImpl processEngineConfiguration;
+
+  /**
+   * Scenario can't be successfully run without modifying runtime code
+   * since we only use H2 in the Spring integration tests, and this test
+   * requires the hard-coded CRDB check to work.
+   */
```

Comment:
Can we at least write a unit test directly for the SpringTransactionInterceptor class? I.e. mock the transaction manager interface and then test the logic?

Same for JTA.

Reply:
The problem is that the "concurrency error" detection logic only kicks in if we use CRDB. 

I'm not sure what we would achieve by mocking the transaction manager interface since the same CRDB validation would exist. One idea would be to mock the transaction manager interface and test the `DbSqlSession.isCrdbConcurrencyConflictOnCommit(...)` method.

Reply:
Why not mock the engine configuration as well? Not great, but better than no test.

Reply:
@ThorbenLindhauer have a look at the last commit (a27fbc6). I temporarily swap out the `databaseType` property to test out the `SpringTransactionInterceptor` logic. The Spring Integration test works like this.

I tried to implement something similar for JTA in the Process Engine test suite without success. We test the `JtaProcessEngineConfiguration` in the QA `integration-tests-engine` module. However, we only use Postgres in the related jobs. Do you think I should try to implement a test there?